### ISSUE TITLE
ReservedThreadExecutorTest.testShrink is a flappy test

### ISSUE THREAD

**Issue by joakime** [eclipse/jetty.project#2211]
CI has reported this test as failing.

It only seems to fail on Java 9 (has not yet failed on Java 8)

It is possible to reproduce locally.
Just run this test in repeat until failure.

On my system, it fails at approx the 4th iteration.
I'm using Java 9.0.4 when testing.

```
java.lang.AssertionError:
Expected: a value less than <1>
but: <1> was equal to <1>
at org.eclipse.jetty.util.thread.ReservedThreadExecutorTest.testShrink(ReservedThreadExecutorTest.java:178)
```

**Comment by joakime** [comment]
DEBUG log on a failed execution ...

```
2018-02-21 14:37:57.497:DBUG:oejut.ReservedThreadExecutor:main: ReservedThreadExecutor@6b1274d2{s=0/2,p=0}
2018-02-21 14:37:57.997:DBUG:oejut.ReservedThreadExecutor:main: ReservedThreadExecutor@6b1274d2{s=0/2,p=0} tryExecute NOOP!
2018-02-21 14:37:57.997:DBUG:oejut.ReservedThreadExecutor:main: ReservedThreadExecutor@6b1274d2{s=0/2,p=1} startReservedThread p=1
2018-02-21 14:37:57.998:DBUG:oejut.ReservedThreadExecutor:Thread-15: ReservedThreadExecutor@6b1274d2{s=1/2,p=1}@4bdfc517 started
2018-02-21 14:37:57.998:DBUG:oejut.ReservedThreadExecutor:Thread-15: ReservedThreadExecutor@6b1274d2{s=1/2,p=0}@4bdfc517 waiting
2018-02-21 14:37:58.498:DBUG:oejut.ReservedThreadExecutor:main: ReservedThreadExecutor@6b1274d2{s=1/2,p=0} tryExecute NOOP!
2018-02-21 14:37:58.498:DBUG:oejut.ReservedThreadExecutor:main: ReservedThreadExecutor@6b1274d2{s=0/2,p=0}@4bdfc517 offer NOOP!
2018-02-21 14:37:58.498:DBUG:oejut.ReservedThreadExecutor:main: ReservedThreadExecutor@6b1274d2{s=0/2,p=1} startReservedThread p=1
2018-02-21 14:37:58.498:DBUG:oejut.ReservedThreadExecutor:Thread-15: ReservedThreadExecutor@6b1274d2{s=0/2,p=1}@4bdfc517 task=NOOP!
2018-02-21 14:37:58.498:DBUG:oejut.ReservedThreadExecutor:Thread-15: ReservedThreadExecutor@6b1274d2{s=1/2,p=1}@4bdfc517 waiting
2018-02-21 14:37:58.499:DBUG:oejut.ReservedThreadExecutor:Thread-16: ReservedThreadExecutor@6b1274d2{s=2/2,p=1}@3547d2b4 started
2018-02-21 14:37:58.499:DBUG:oejut.ReservedThreadExecutor:Thread-16: ReservedThreadExecutor@6b1274d2{s=2/2,p=0}@3547d2b4 waiting
2018-02-21 14:37:58.999:DBUG:oejut.ReservedThreadExecutor:main: ReservedThreadExecutor@6b1274d2{s=2/2,p=0} tryExecute NOOP!
2018-02-21 14:37:58.999:DBUG:oejut.ReservedThreadExecutor:main: ReservedThreadExecutor@6b1274d2{s=1/2,p=0}@3547d2b4 offer NOOP!
2018-02-21 14:37:58.999:DBUG:oejut.ReservedThreadExecutor:Thread-16: ReservedThreadExecutor@6b1274d2{s=1/2,p=0}@3547d2b4 task=NOOP!
2018-02-21 14:37:58.999:DBUG:oejut.ReservedThreadExecutor:Thread-16: ReservedThreadExecutor@6b1274d2{s=2/2,p=0}@3547d2b4 waiting
2018-02-21 14:37:59.499:DBUG:oejut.ReservedThreadExecutor:main: ReservedThreadExecutor@6b1274d2{s=2/2,p=0} tryExecute NOOP!
2018-02-21 14:37:59.499:DBUG:oejut.ReservedThreadExecutor:main: ReservedThreadExecutor@6b1274d2{s=1/2,p=0}@3547d2b4 offer NOOP!
2018-02-21 14:37:59.499:DBUG:oejut.ReservedThreadExecutor:Thread-16: ReservedThreadExecutor@6b1274d2{s=1/2,p=0}@3547d2b4 task=NOOP!
2018-02-21 14:37:59.500:DBUG:oejut.ReservedThreadExecutor:Thread-16: ReservedThreadExecutor@6b1274d2{s=2/2,p=0}@3547d2b4 waiting
2018-02-21 14:37:59.500:DBUG:oejut.ReservedThreadExecutor:Thread-15: ReservedThreadExecutor@6b1274d2{s=2/2,p=0}@4bdfc517 IDLE
2018-02-21 14:37:59.500:DBUG:oejut.ReservedThreadExecutor:Thread-15: ReservedThreadExecutor@6b1274d2{s=2/2,p=0} tryExecute STOP!
2018-02-21 14:37:59.500:DBUG:oejut.ReservedThreadExecutor:Thread-15: ReservedThreadExecutor@6b1274d2{s=1/2,p=0}@3547d2b4 offer STOP!
2018-02-21 14:37:59.500:DBUG:oejut.ReservedThreadExecutor:Thread-16: ReservedThreadExecutor@6b1274d2{s=1/2,p=0}@3547d2b4 task=STOP!
2018-02-21 14:37:59.500:DBUG:oejut.ReservedThreadExecutor:Thread-16: ReservedThreadExecutor@6b1274d2{s=1/2,p=0}@3547d2b4 Exited
2018-02-21 14:37:59.999:DBUG:oejut.ReservedThreadExecutor:main: ReservedThreadExecutor@6b1274d2{s=1/2,p=0} tryExecute NOOP!
2018-02-21 14:37:59.999:DBUG:oejut.ReservedThreadExecutor:main: ReservedThreadExecutor@6b1274d2{s=0/2,p=0}@4bdfc517 offer NOOP!
2018-02-21 14:37:59.999:DBUG:oejut.ReservedThreadExecutor:main: ReservedThreadExecutor@6b1274d2{s=0/2,p=1} startReservedThread p=1
2018-02-21 14:37:59.999:DBUG:oejut.ReservedThreadExecutor:Thread-15: ReservedThreadExecutor@6b1274d2{s=0/2,p=1}@4bdfc517 task=NOOP!
2018-02-21 14:38:00.000:DBUG:oejut.ReservedThreadExecutor:Thread-15: ReservedThreadExecutor@6b1274d2{s=1/2,p=1}@4bdfc517 waiting
2018-02-21 14:38:00.000:DBUG:oejut.ReservedThreadExecutor:Thread-17: ReservedThreadExecutor@6b1274d2{s=2/2,p=1}@29cf02b0 started
2018-02-21 14:38:00.000:DBUG:oejut.ReservedThreadExecutor:Thread-17: ReservedThreadExecutor@6b1274d2{s=2/2,p=0}@29cf02b0 waiting
2018-02-21 14:38:01.000:DBUG:oejut.ReservedThreadExecutor:Thread-15: ReservedThreadExecutor@6b1274d2{s=2/2,p=0}@4bdfc517 IDLE
2018-02-21 14:38:01.000:DBUG:oejut.ReservedThreadExecutor:Thread-15: ReservedThreadExecutor@6b1274d2{s=2/2,p=0} tryExecute STOP!
2018-02-21 14:38:01.000:DBUG:oejut.ReservedThreadExecutor:Thread-15: ReservedThreadExecutor@6b1274d2{s=1/2,p=0}@29cf02b0 offer STOP!
2018-02-21 14:38:01.000:DBUG:oejut.ReservedThreadExecutor:Thread-17: ReservedThreadExecutor@6b1274d2{s=1/2,p=0}@29cf02b0 task=STOP!
2018-02-21 14:38:01.000:DBUG:oejut.ReservedThreadExecutor:Thread-17: ReservedThreadExecutor@6b1274d2{s=1/2,p=0}@29cf02b0 Exited
2018-02-21 14:38:02.001:DBUG:oejut.ReservedThreadExecutor:Thread-15: ReservedThreadExecutor@6b1274d2{s=1/2,p=0}@4bdfc517 IDLE
2018-02-21 14:38:02.001:DBUG:oejut.ReservedThreadExecutor:Thread-15: ReservedThreadExecutor@6b1274d2{s=1/2,p=0} tryExecute STOP!
2018-02-21 14:38:02.001:DBUG:oejut.ReservedThreadExecutor:Thread-15: ReservedThreadExecutor@6b1274d2{s=0/2,p=0}@4bdfc517 offer STOP!
2018-02-21 14:38:02.001:DBUG:oejut.ReservedThreadExecutor:Thread-15: ReservedThreadExecutor@6b1274d2{s=0/2,p=0}@4bdfc517 task=STOP!
2018-02-21 14:38:02.001:DBUG:oejut.ReservedThreadExecutor:Thread-15: ReservedThreadExecutor@6b1274d2{s=0/2,p=0}@4bdfc517 Exited

java.lang.AssertionError: 
Expected: is <0>
     but: was <1>
Expected :is <0>
     
Actual   :<1>

	at org.eclipse.jetty.util.thread.ReservedThreadExecutorTest.testShrink(ReservedThreadExecutorTest.java:178)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:67)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

**Comment by joakime** [comment]
Remarkably difficult to debug.

The `_reservedExecutor.getAvailable()` returns `1`, but by the time debug kicks in the `_size.get()` is `0`

**Comment by gregw** [comment]
Fixed by adcb2762fed5d0cab9158887f379945b22c0df20
